ad/views.py

```diff
@@ -82,4 +83,62 @@ def get_model(self):
     
     def get_queryset(self):
         model = self.get_model()
-        return model.objects.all()
\ No newline at end of file
+        return model.objects.all()
+    
+class AdUpdateView(LoginRequiredMixin, UpdateView, ModelMappingMixin):
+    template_name = 'ad/ad_form.html'
+    context_object_name = 'ad'
+    
+    def get_model(self):
+        adtype = self.kwargs['adtype']
+        model = self.model_mapping.get(adtype)
+        if not model:
+            raise Http404(f"No model found for ad type: {adtype}")
+        return model
+    
+    def get_form_class(self):
+        adtype = self.kwargs['adtype']
+        return self.form_mapping.get(adtype)
+    
+    def get_queryset(self):
+        model = self.get_model()
+        return model.objects.all()
```

Comment:
Extract common methods used in edit view and update view to ModelMappingMixin to avoid repetition.